bgpd/bgp_rpki.c

```diff
@@ -371,26 +371,36 @@ static int bgpd_sync_callback(struct thread *thread)
 	afi_t afi = (rec.prefix.ver == LRTR_IPV4) ? AFI_IP : AFI_IP6;
 
 	for (ALL_LIST_ELEMENTS_RO(bm->bgp, node, bgp)) {
-		safi_t safi;
+		struct peer *peer;
+		struct listnode *peer_listnode;
 
-		for (safi = SAFI_UNICAST; safi < SAFI_MAX; safi++) {
-			if (!bgp->rib[afi][safi])
-				continue;
+		for (ALL_LIST_ELEMENTS_RO(bgp->peer, peer_listnode, peer)) {
```

Comment:
is this really what we want to do?

Suppose I have 200 peers.  You'll look at the same bgp->rib 200+ times.  I do not understand the issue here as to why peer->bgp->rib is different than the bgp->rib... what is the subtlety I am missing?

Reply:
I noticed routes vanishing when revalidation occurred, sometimes in the hundreds of thousands sometimes only a few.
Triggering soft reconfiguration changes would bring them all back.
So I looked at the soft reconfig code and the only relevant difference I found was that I used bgp->rib and soft reconfig uses peer->bgp->rib.

After giving it some thought I came to the conclusion that using bgp->rip is in this case wrong, because only routes that are currently selected would be revalidated. Unless I'm misunderstanding what bgp_update does.